### DISCUSSION TITLE
Why does Puma::DSL not implement `#on_stopped`? Possible oversight?

### DISCUSSION THREAD

**Discussion by emilyst** [puma/puma#3379] [Q&A] [answered]
Hi, maybe someone here will know this, but I can't find the answer after searching.

`Puma::Events` defines three hooks for the lifecycle of a server: `#on_booted`, `#on_restart`, and `#on_stopped`. Each takes a block to execute when this event occurs.

In `Puma::DSL`, there's an implementation of `#on_booted` and `#on_restart`, but nothing for `#on_stopped`. (Curiously, the two existing event hook methods are slightly different in implementation, but I didn't dig further to figure out why.)

Additionally, [the `README` itself refers prominently to `on_stopped`](https://github.com/puma/puma?tab=readme-ov-file#master-process-lifecycle-hooks). But of course that method does not exist in the DSL, so you get the following error:

> config/puma.rb:45:in `_load_from': undefined method `on_stopped' for an instance of Puma::DSL (NoMethodError)

Is this possibly just an oversight of some kind? Or maybe the method was removed for whatever reason, but the documentation remained in place?

The reason I'm interested is that, where I work, we run Puma in single mode for now, and we want to flush Kafka events from the producer's buffer when the process receives `SIGTERM` before stopping. Otherwise, we may drop buffered messages on the floor. Right now, our options seem to be:

1. Write a plugin which can get access to `Puma::Events` (maybe overkill for a one-liner?).
2. Write some code in `config/puma.rb` which digs the object out of the DSL configuration (e.g., `instance_var_get(:@config).events.on_stopped`).
3. Switch to clustered mode—again, a bit silly to do just for this one thing.
4. Write our own `SIGTERM` handler, independent of Puma.

I'm not sure which option is best. Any thoughts?

Thanks in advance!

**Comment by dentarg** [comment]
It looks like @johnnyshields made it up in https://github.com/puma/puma/pull/3363 and I did not catch it!

**Reply by johnnyshields** [reply]
It is defined in lib/puma/events.rb. Guess it is not exposed by DSL (oops). We should probably expose it however…

**Reply by emilyst** [reply]
If you're supportive of implementing it, check out this pull request I've added: https://github.com/puma/puma/pull/3380.

**Comment by dentarg** [comment, marked as answer]
Addressed by https://github.com/puma/puma/pull/3411
